Docs : Add Bootstrap scrollspy

medubelko opened the pull request:
In an effort to improve UX on the docs after I removed the in-article indexes, I've put together a draft scrollspy implementation based on Bootstrap. The intention is to provide signposting and a mini-nav for article contents. For comparison: [Bootstrap](https://getbootstrap.com/docs/3.4/javascript/), [Microsoft](https://docs.microsoft.com/en-us/powershell/scripting/getting-started/starting-windows-powershell?view=powershell-6), and [IBM](https://cloud.ibm.com/docs/bare-metal?topic=bare-metal-about-bm#about-bm).

Obviously it won't work for every article, so I've limited it to articles that have more than 3 headings or are taller than 1600px. It's also only visible on desktop.

Short article preview:
```
~michaeldu/dev/gafferDocsBuilds/docsScrollspy/html/WorkingWithScenes/AnatomyOfAScene/index.html
```

Long article preview:
```
~michaeldu/dev/gafferDocsBuilds/docsScrollspy/html/WorkingWithTheNodeGraph/Box/index.html
```

Items I'd like feedback on:
- Should we keep the conditional presence based on article length? It creates an inconsistency, but to me having the scrollspy appear in articles that have 2 headings or 10 words seems silly. In such cases the links wouldn't be useful, anyway.
- Should it show H3-H6s at all? In some articles it's fine, but in others it makes the scrollspy quite long. Some of the reference articles (`~michaeldu/dev/gafferDocsBuilds/docsScrollspy/html/Reference/ScriptingReference/CommonOperations/index.html`) have a series of H3s in a row that are all a sentence or two.
    - If we keep them in, one way to improve them would be to make the scrollspy a collapsible [accordion](https://en.wikipedia.org/wiki/Accordion_(GUI)). My main reason against doing this was that it hampers the signposting and duplicates the behaviour of the main nav. It would take a bit more effort, but it's doable.
    - Another would be to ignore any individual sections shorter than a certain length, but I can imagine that would seem fickle and confusing.
- Scrollspies don't work on smaller devices. Should I put the effort into moving the container under the article title on mobile, and make it a static index? If I did this, I would have to add accordion functionality; the container would usually be too long otherwise.
- How frustrating is it that short sections at the bottoms of articles can't properly be spied/linked (c.f. the end of _Anatomy of a Scene_)? I can solve this by adding a chunk of negative space to the article bottoms.
- @themissingcow How does the contrast look on a retina monitor?

medubelko commented:
Argh, it looks like later Sphinx has changed the placement of the links in the DOM. I'll have to add a fallback check.

medubelko commented:
@themissingcow Try giving it a whirl now.

themissingcow commented:
> @themissingcow Try giving it a whirl now.

:+1: Looks great!

themissingcow commented:
Think this is a great addition, good idea.

> * Should we keep the conditional presence based on article length? It creates an inconsistency, but to me having the scrollspy appear in articles that have 2 headings or 10 words seems silly. In such cases the links wouldn't be useful, anyway.

Maybe it'd be good to keep it in all the time, I tried it on "Launching Gaffer for the First Time" and there it would be useful, as the OSX section is off-screen. I guess its harder to make the hiding rules work universally?  You just ignore it if its not relevant? Unless it could be a curated thing, such that you put some markup on the pages you don't want it on?

> * Should it show H3-H6s at all? In some articles it's fine, but in others it makes the scrollspy quite long. 

Perhaps not, as there is no scrolling in scrollspy itself, so it often goes off screen... 

>   * If we keep them in, one way to improve them would be to make the scrollspy a collapsible [accordion](https://en.wikipedia.org/wiki/Accordion_(GUI)). My main reason against doing this was that it hampers the signposting and duplicates the behaviour of the main nav. It would take a bit more effort, but it's doable.

Don't know if its possible, but what about if it didn't show H3s until you were on the parent H2?  
I guess if you take the 'Working with the Box node' page as an example, 'Entering and exiting a Box' is only a screen full, but 'Promoting and demoting a plug' would benefit from them...

>   * Another would be to ignore any individual sections shorter than a certain length, but I can imagine that would seem fickle and confusing.

Yeah, it'd be weird to figure 'why isn't that one showing up'...

> * Scrollspies don't work on smaller devices. Should I put the effort into moving the container under the article title on mobile, and make it a static index? If I did this, I would have to add accordion functionality; the container would usually be too long otherwise.

Do we have usage stats? Are there many people on mobile? I don't know how much use it'd be unless it was available via a popover. If its at the top, its only useful for an initial jump... and then it'd be hard to make it fit. Seems not worth the effort to me unless we have majority on mobile for the actual docs?

> * How frustrating is it that short sections at the bottoms of articles can't properly be spied/linked (c.f. the end of _Anatomy of a Scene_)? I can solve this by adding a chunk of negative space to the article bottoms.

Do you mean that it scrolls, but the clicked title isn't at the top? That doesn't bother me too much. Makes sense as its the bottom. I think having all the dead space at the bottom would be more unusual... You'd have to add a lot of space to accommodate for all screens....

> * @themissingcow How does the contrast look on a retina monitor?

I don't know if its more a gamma issue, but I think the 'current' heading could be a little more prominent. Gaffer red? The mid-gray of the index seems darker than on linux.

medubelko commented:
@themissingcow Thanks for the detailed review.

> Unless it could be a curated thing, such that you put some markup on the pages you don't want it on?

I like this idea. It gives us more control without running into unexpected edge cases.

> there is no scrolling in scrollspy itself, so it often goes off screen... Don't know if its possible, but what about if it didn't show H3s until you were on the parent H2?...

I was originally planning on having it scrollable. I realise in hindsight how it must have been clipped for most articles on smaller laptop screens. Maybe making it accordion is unavoidable, given how long some of the articles are. Now that I think about it, having it be a right-side element and scrollable is really dangerous, because that's the region where the user's mouse is likely to be to scroll the main article.

> Seems not worth the effort to me unless we have majority on mobile for the actual docs?

About 11%, according to recent stats. You're right.

> The mid-gray of the index seems darker than on linux.

I'll see about subtly increasing the contrast.

medubelko commented:
@themissingcow Give it a try now. Here's what I've changed:

- [x] Removed exclusion based on article length detection.
- [x] Added detection for a user token (`!NOSCROLLSPY`) in an HTML comment as the first line, that will disable the Scrollspy on a per-article basis.
- [x] Compressed sub-headings under H2s.
- [x] Added scrolling to the scrollspy itself.
- [x] Increased contrast between active and inactive headers.

themissingcow commented:
Hey, thanks for those updates.

> * [x]  Removed exclusion based on article length detection.
> * [x]  Added detection for a user token (`!NOSCROLLSPY`) in an HTML comment as the first line, that will disable the Scrollspy on a per-article basis.

Great, nice one.

> * [x]  Compressed sub-headings under H2s.

This works I think, but I noticed that we now have `h4` tags appearing in the index, so even with collapsing the non-current sections the index can be really long. Might be more readable if we stopped at `h3`?
 
> * [x]  Added scrolling to the scrollspy itself.

2 minds on this one, it does  make it  visually much more intrusive:

![image](https://user-images.githubusercontent.com/896779/59342181-282c8200-8d01-11e9-8b9f-a716efc50d44.png)

I wonder if it'd be as much of an issue though if we didn't include beyond `h3` as it might fit on most of the time?

> * [x]  Increased contrast between active and inactive headers.

That's better, thanks.

medubelko commented:
> I wonder if it'd be as much of an issue though if we didn't include beyond h3 as it might fit on most of the time?

We're always going to produce articles that are long enough to cause the scroll to appear. Gaffer is complicated, and tutorials will be necessarily long. In the copy I'm generally trying to _avoid_ H4s/H5s in favour of H2s/H3s, because it reduces mental load (there's a common assumption that more hierarchy = better, because added structure feels like it should communicate patterns and relationships, but 2-3 levels are all that readers can really handle in online documents... but I digress).

So, it's the scroll itself that's the issue. Just for comparison, here's how Docker's scrollspy appears on a very long article: https://docs.docker.com/engine/reference/commandline/run/. I suspect ours is disagreeable because it harshly breaks the right side of the page. I'll make it right-align and thinner and more compact, as a test.

medubelko commented:
@themissingcow Here's what I've addressed in these latest commits.

Changes to reduce footprint on screen:
- [x] Shrink text size slightly (14px → 13px).
- [x] Right-align scrollspy.
- [x] Remove double-bisection of right edge of article; only the scrollspy has the shadow now.
- [x] Make scrollspy take up full window height, so it doesn't feel popped-out.

Unrelated changes:
- [x] Preserve text formatting in headings (accidentally undone in previous commits).
- [x] Removed `↑Top` button, for feedback.

The scrollbar is now flush with the browser scrollbar. Let me know what you think.

themissingcow commented:
> . Just for comparison, here's how Docker's scrollspy appears on a very long article: 

![docker](https://user-images.githubusercontent.com/896779/59665540-04f34e00-91ab-11e9-8d50-fea11c206206.png)

To me, that's even worse, as the scrollspy scroll bar dominates the actual page scroll bar, which I guess just infers that very few people actually use scroll handles these days and probably use wheel/gestures? If I was going to use it, I'd grab that big one, not the tiny one on the top right.

themissingcow commented:
Hey @medubelko, think this looks really good now. Nice one!

> * [x]  Right-align scrollspy.

Still appears left for me, unless I misunderstand you. But think it looks great how it is here:

![scrollspy](https://user-images.githubusercontent.com/896779/59676077-a1731b80-91be-11e9-92ac-3c515d48e4a8.png)

> The scrollbar is now flush with the browser scrollbar. Let me know what you think.

Personally, find it the same as the Docker site - in that the scrollspy scrollbar eclipses the real scroll bar, but I dont think there is really a practical solution, and its at least consistent with other sites and now we don't show the scroll bar that often...  so LGTM.

I guess the only thing remaining would be to make sure we have scrollspy off for some of the shorter pages, generally I guess that's the top level pages in each section, listed at the top level of the left-hand nav. Feels like something is awry when you have "In this article" with nothing underneath it:

![scrollspy2](https://user-images.githubusercontent.com/896779/59676343-47268a80-91bf-11e9-96af-02b0e8b58ba6.png)

Adding scrollspy is a great improvement though :+1:! Thanks.

medubelko commented:
> Still appears left for me

I meant the whole container itself. :)

> I guess the only thing remaining would be to make sure we have scrollspy off for some of the shorter pages

Will do, now that you've signed off on the look + position.

medubelko commented:
Latest changes:
- [x] Add `!NO_SCROLLSPY` tag to indexes and short articles.
- [x] Do a final style pass.
- [x] Rebase.
